### PR TITLE
New explanation about template universe polymorphism

### PR DESCRIPTION
**Pull request by lephe**
Discussed on Zulip: https://coq.zulipchat.com/#narrow/stream/437203-Coq-Platform-docs/topic/Template.20vs.2E.20proper.20universe.20polymorphism.20tutorial

The claimed scope, for now, is just an explanation of what template polymorphism does and doesn't do, with bridges around it to where the problem it solves come from (as explained with monomorphic universes) and how the limitations can be worked around (with universe polymorphism).

I think a detailed explanation of universes would be mostly independent so probably better separated. If such an explanation is written, linking to it could replace the current first section.

On the other hand, discussing other limitations of template universe polymorphism and better showing the tricks that can be used to hack around them would be, IMO, clearly in scope for this document. I could see a generalization to "more things about template polymorphism" more than to "more things about universes". The current text is still right at the limits of my understanding so I'm not really confident I can extend further.

Cc @YaZko

### CODE REVIEW

File: src/Explanation_Template_Polymorphism.v
Comment (SkySkimmer) [inline src/Explanation_Template_Polymorphism.v:64]: I wouldn't say that infinity is what gives consistency.

File: src/Explanation_Template_Polymorphism.v
Comment (SkySkimmer) [inline src/Explanation_Template_Polymorphism.v:136]: in some places you use Check and others Type (the command), probably you should pick one

File: src/Explanation_Template_Polymorphism.v
Comment (lephe) [reply src/Explanation_Template_Polymorphism.v:64]: Any suggestions for a more correct statement here? I'd just like to give closure on how this avoids Girard's paradox.

File: src/Explanation_Template_Polymorphism.v
Comment (lephe) [reply src/Explanation_Template_Polymorphism.v]: Good catch. I went for `Check` to reduce lexical pressure.

File: src/Explanation_Template_Polymorphism.v
Comment (SkySkimmer) [reply src/Explanation_Template_Polymorphism.v:64]: I think the order being well founded is what gives consistency.

**Comment by lephe** [comment]
Gentle ping. I added the main contributors field from #36. Is there anything else I should do?

**Comment by thomas-lamiaux** [comment]
Hi @lephe , it is going a bit slow because of the holiday. Considering the discussions we had on the zulip, I suggest the following steps: 
1. I review it
2. after changes, we merge it without waiting for completion with further content
3. we create issues "intro tuto" and "complete the trade of" ? 

Does that sound good to you ?

**Comment by lephe** [comment]
Ah, perfect, thanks for the process. This sounds good to me.

File: src/Explanation_Template_Polymorphism.v
Comment (thomas-lamiaux) [inline src/Explanation_Template_Polymorphism.v:238]:
```suggestion
Fail Definition mprod_lazy {A B: Type} (a: A) (b: B) : mprod (lazyT A) (lazyT B) :=
```

File: src/Explanation_Template_Polymorphism.v
Comment (thomas-lamiaux) [inline src/Explanation_Template_Polymorphism.v:266]: Complete with a wheras ...

File: src/Explanation_Template_Polymorphism.v
Comment (thomas-lamiaux) [inline src/Explanation_Template_Polymorphism.v:268]: Not sure what is the point of this comment. 
Plus Prop is complicated.

File: src/Explanation_Template_Polymorphism.v
Comment (thomas-lamiaux) [inline src/Explanation_Template_Polymorphism.v:299]: Sounds like you had a lot of issues with it :laughing:

File: src/Explanation_Template_Polymorphism.v
Comment (thomas-lamiaux) [inline src/Explanation_Template_Polymorphism.v:333]: From your explanations, it is sounds a bit like template polymorphism is a fake solution and a bit pointless :sweat_smile:

File: src/Explanation_Template_Polymorphism.v
Comment (thomas-lamiaux) [inline src/Explanation_Template_Polymorphism.v:372]: It sounds a bit bitter

File: src/Explanation_Template_Polymorphism.v
Comment (thomas-lamiaux) [inline src/Explanation_Template_Polymorphism.v:315]: Should there be examples ?

**Review by thomas-lamiaux — CHANGES_REQUESTED** [review]
I have looked at the tutorials, it is very interesting and a very good version on the subject. Thanks for the contribution. 

Concerning the tutorial and the structure, I suggest to:  
- Section is unclear and most of section 2 is an introduction to universes as
indicated by "Let's see this in action."
  Hence, I suggest moving / merging all such content with Section 1.
  Like everything which is not related to mprod and the limitations of using it.
  Then dedicate section 2 to monomorphic and limitations
- Maybe add subsections in section 3 and 4. 
  Like "Principles of (Template Poly // UnivPoly) / "advantages" / "limitations". 
  I think it will make a bit easier to read and easier to improve later on, 
  for a very little cost
- It is not clear in section 2 what monomorphic means, maybe elaborate on it 
- Maybe rework the intro to say this tuto discuss the two approaches, their advantages and limitations as it is the end goal. In which case, maybe mention it is a first version or sth ? 

Otherwise sounds great !

File: src/Explanation_Template_Polymorphism.v
Comment (SkySkimmer) [reply src/Explanation_Template_Polymorphism.v:117]: All global universes are > Set.
Records of why this was decided are lost but AFAICT the main consequence is that they are prevented from being made = Set, which seems desired as if the user wanted a universe to be = Set they should have written Set not Type.

**Comment by lephe** [comment]
Thanks, I'll go through that!

One point on which I need an expert's opinion is the status of template to full universe polymorphism. The latter is more expressive on paper, and I thought it was seen as superseding template polymorphism. I'd heard in particular of attempts to write a universe polymorphic standard library, suggesting that this was desirable.

I indeed had issues with it. ^^" At some point in my last project I debugged universes for 20-30 hours, with some problems arising from the project itself, the others from template polymorphism in the standard library. Having to untangle these and sometimes purge standard types from sections was very painful (hence the slightly bad vibe, sorry—will fix). As of now this project still reimplements the product type btw, so I was truly writing under the assumption that template polymorphism was insufficient.

Can anyone more involved with this stuff comment on how the relationship between these two types of polymorphism should be presented? Is template polymorphism really considered to be superseded by full universe polymorphism (either now or in the future)? Are there contexts in which it is specifically desirable to use one type of polymorphism over the other? Maybe some context about what the project's direction is?

**Comment by thomas-lamiaux** [comment]
Not sure if it answers, no time to read it again but I am recalling https://coq.zulipchat.com/#narrow/stream/437203-Coq-Platform-docs/topic/Template.20vs.2E.20proper.20universe.20polymorphism.20tutorial/near/454897533

File: src/Explanation_Template_Polymorphism.v
Comment (lephe) [reply src/Explanation_Template_Polymorphism.v:268]: Well, more examples that the universe of `T * T` depends on the universe of `T`.

File: src/Explanation_Template_Polymorphism.v
Comment (lephe) [reply src/Explanation_Template_Polymorphism.v:315]: I'm not really familiar with these, I just know they exist.

**Comment by lephe** [comment]
Pushed things. I've addressed all comments so far except for:

> Maybe rework the intro to say this tuto discuss the two approaches, their advantages and limitations as it is the end goal. In which case, maybe mention it is a first version or sth ?

Hoping for more background, the post linked above goes into useful technical details but I'm still not sure whether template polymorphism is supposed to be superseded and used for legacy reasons + cases where univ poly blows up or whether these are really two parallel systems that are designed to exist concurrently.

> Explain somewhere that whenever Type is used, then a new variable is created or sth and the universe constrains. Maybe add a Print Universes Subgraph (lazyT.u0 lazyT.u1). 

I'll have to look up what happens exactly as I don't know directly.

> Should there be examples [of tricks to bypass limitations of template universe polymorphism] ?

I also don't know these very well. I just personally inline stuff. Gaëtan recently demonstrated how [eta-expansion helps](https://coq.zulipchat.com/#narrow/stream/437203-Coq-Platform-docs/topic/Template.20vs.2E.20proper.20universe.20polymorphism.20tutorial/near/454844195) but it seems like the trick should be an entire section by itself. Which would be on topic, IMO.

File: src/Explanation_Template_Polymorphism.v
Comment (thomas-lamiaux) [inline src/Explanation_Template_Polymorphism.v:173]: What is it ?

File: src/Explanation_Template_Polymorphism.v
Comment (thomas-lamiaux) [inline src/Explanation_Template_Polymorphism.v:221]: I think it would be good here to explain a bit much how each constrains appear because it easy to get lost in the direction of the inequalities

File: src/Explanation_Template_Polymorphism.v
Comment (thomas-lamiaux) [inline src/Explanation_Template_Polymorphism.v:228]: That is an example of how the issue arises, but it is not a very general comment.
Would you have a more general explanation for when this issues arise to add in complement just before ?

File: src/Explanation_Template_Polymorphism.v
Comment (thomas-lamiaux) [inline src/Explanation_Template_Polymorphism.v:263]: About prod is printed but there are no explanations, they come further. 
I suggest adding a line or two to discuss the change here.

File: src/Explanation_Template_Polymorphism.v
Comment (thomas-lamiaux) [inline src/Explanation_Template_Polymorphism.v:323]: Could you develop here, I have some trouble understanding what is the issue as the type of state is 
```
prod : Type@{prod.u0} -> Type@{prod.u1} -> Type@{max(prod.u0,prod.u1)}
state : Type@{state.u0} -> Type@{state.u1} -> Type@{max(state.u0,state.u1)}
```
which to me looks like the same as `prod` and not `mprod`

**Review by thomas-lamiaux — COMMENTED** [review]
structure looks better

File: src/Explanation_Template_Polymorphism.v
Comment (thomas-lamiaux) [reply src/Explanation_Template_Polymorphism.v:331]: I suggest creating a subsection as you mentioned and leave it for further contributions

**Comment by thomas-lamiaux** [comment]
Moving further, I suggest fixing the last tiny issues, then merge it.
It does not matter if it not perfect

**Comment by thomas-lamiaux** [comment]
@lephe Have you had time to progress ? I would like to merge it soon

**Comment by lephe** [comment]
I can do the next (supposedly last) pass tomorrow.

**Comment by thomas-lamiaux** [comment]
Ok sounds good to me. I'll merge it and then create issues to improve it

File: src/Explanation_Template_Polymorphism.v
Comment (lephe) [reply src/Explanation_Template_Polymorphism.v:323]: The difference is not in the type, it's in the annotation "X is template universe polymorphic" shown by  `About`. This status affects the way instances of the definition are treated, and that's what makes the difference. I clarified.

**Comment by lephe** [comment]
@thomas-lamiaux Ready to merge if nothing major remains.

**Pull request merged**
